lib/ansible/playbook/play.py

```diff
@@ -293,8 +293,18 @@ def compile(self):
         t.args['_raw_params'] = 'flush_handlers'
         t.implicit = True
         t.set_loader(self._loader)
-        t.tags = self.tags
-        if t.evaluate_tags(self.only_tags, self.skip_tags, all_vars=self.vars):
+
+        if self.tags:
+            # Avoid calling flush_handlers in case the whole play is skipped on tags,
+            # this could be performance improvement since calling flush_handlers on
+            # large inventories could be expensive even if no hosts are notified
+            # since we call flush_handlers per host.
+            # Block.filter_tagged_tasks ignores evaluating tags on implicit meta
+            # tasks so we need to explicitly call Task.evaluate_tags here.
+            t.tags = self.tags
+            if t.evaluate_tags(self.only_tags, self.skip_tags, all_vars=self.vars):
+                flush_block.block = [t]
+        else:
```

Comment:
i would do here the same as in gather_facts implicit task, assign 'inherited tags' or 'always' tag, then always evaluate tags

Reply:
it is just for corner cases as you can do `--skip-tags always`  or similar to handle the fact gathering currently, I thought it might make sense to make all 'implicit' tasks behave the same.

Reply:
I feel like assigning the `always` tag to the implicit `flush_handlers` tasks to allow `--skip-tags always` to skip them would be a surprising change in behavior at this point, no?

My intent for this was to just not run (include into the play) implicit `flush_handlers` in case the whole play is skipped on tags.

Reply:
true, but we should note in docs that they are not skippable for those cases and that they 'always' run unless otherwise tagged

Reply:
I am not sure we should mention implicit `flush_handlers` in docs. It seems that mentioning that handlers themselves do not support tags (https://github.com/ansible/ansible-documentation/pull/1900) suffice. This PR just fixes what is expected that when play is tagged and it is skipped on its tags then nothing from that play runs. Which isn't true in `devel` where we still run `flush_handlers` needlessly since nothing could possibly notify any handler. So all this is transparent to the user and they shouldn't have to know about any of this; just that handlers run after `pre_tasks`/`roles+tasks`/`post_tasks` which do already document.

Reply:
since no one has complained yet ... safe bet